### ISSUE TITLE
Cannot create new Mailbox or set Password 

### ISSUE THREAD

**Issue by huddx01** [opensolutions/ViMbAdmin#277]
Hi,

I cannot create new Mailboxes or set another password for the mailboxes since i modified something in the config...

It seems that the "doeveadm pw" fails... but if i run it as root, it works... 

Example:

`/usr/bin/doveadm pw -s 'SHA512-CRYPT' -u 'xxx@mydomain.com' -p 'myPass12345678!'
{SHA512-CRYPT}$6$oLZ5VcgQoAoYleSu$5B71l9rHAkjkVKyCDFbGAdu6OwkeyYlh00ukniQkwcWZFc3JN8.EqwdhwRC6J5u.NSSuTvVy8pJuLpOPyjIo61`

If i try it wit the www-data privileges, it fails:

`runuser -u www-data -- /usr/bin/doveadm pw -s 'SHA512-CRYPT' -u -u 'xxx@mydomain.com' -p 'myPass12345678!'
doveconf: Fatal: Error in configuration file /etc/dovecot/dovecot.conf line 144: ssl_cert: Can't open file /etc/letsencrypt/live/mydomain.com/fullchain.pem: Permission denied
`

I know it seems to be something with the access rights to the le - certificates... But other apache-vhosts have no problem accessing to it...

Maybe someone has a hint? 

Thanks in advance! 

```
2021-02-18T21:49:26+01:00 DEBUG (7): Logger instantiated
2021-02-18T21:49:26+01:00 DEBUG (7): 

2021-02-18T21:49:26+01:00 DEBUG (7): ErrorController::errorAction()
2021-02-18T21:49:26+01:00 ERR (3): 

************************************************************************
****************************** EXCEPTIONS *******************************
************************************************************************

--------------------------- EXCEPTION --------------------------

Message: Error executing Dovecot password command: /usr/bin/doveadm pw -s 'SHA512-CRYPT' -u xxx@mydomain.com' -p 'myPass12345678!'
Line: 57
File: /srv/vimbadmin/library/ViMbAdmin/Dovecot.php

Trace:

#0 /srv/vimbadmin/library/OSS/Auth/Password.php(97): ViMbAdmin_Dovecot::password()
#1 /srv/vimbadmin/application/controllers/MailboxController.php(316): OSS_Auth_Password::hash()
#2 /srv/vimbadmin/vendor/shardj/zf1-future/library/Zend/Controller/Action.php(516): MailboxController->addAction()
#3 /srv/vimbadmin/vendor/shardj/zf1-future/library/Zend/Controller/Dispatcher/Standard.php(308): Zend_Controller_Action->dispatch()
#4 /srv/vimbadmin/vendor/shardj/zf1-future/library/Zend/Controller/Front.php(954): Zend_Controller_Dispatcher_Standard->dispatch()
#5 /srv/vimbadmin/vendor/shardj/zf1-future/library/Zend/Application/Bootstrap/Bootstrap.php(105): Zend_Controller_Front->dispatch()
#6 /srv/vimbadmin/vendor/shardj/zf1-future/library/Zend/Application.php(384): Zend_Application_Bootstrap_Bootstrap->run()
#7 /srv/vimbadmin/public/index.php(34): Zend_Application->run()
#8 {main}

Array
(
   [0] => Array
       (
           [file] => /srv/vimbadmin/library/OSS/Controller/Trait/Error.php
           [line] => 78
           [function] => compact_debug_backtrace
           [class] => OSS_Debug
       )

   [1] => Array
       (
           [file] => /srv/vimbadmin/vendor/shardj/zf1-future/library/Zend/Controller/Action.php
           [line] => 516
           [function] => errorAction
           [class] => ErrorController
       )

   [2] => Array
       (
           [file] => /srv/vimbadmin/vendor/shardj/zf1-future/library/Zend/Controller/Dispatcher/Standard.php
           [line] => 308
           [function] => dispatch
           [class] => Zend_Controller_Action
       )

   [3] => Array
       (
           [file] => /srv/vimbadmin/vendor/shardj/zf1-future/library/Zend/Controller/Front.php
           [line] => 954
           [function] => dispatch
           [class] => Zend_Controller_Dispatcher_Standard
       )

   [4] => Array
       (
           [file] => /srv/vimbadmin/vendor/shardj/zf1-future/library/Zend/Application/Bootstrap/Bootstrap.php
           [line] => 105
           [function] => dispatch
           [class] => Zend_Controller_Front
       )

   [5] => Array
       (
           [file] => /srv/vimbadmin/vendor/shardj/zf1-future/library/Zend/Application.php
           [line] => 384
           [function] => run
           [class] => Zend_Application_Bootstrap_Bootstrap
       )

   [6] => Array
       (
           [file] => /srv/vimbadmin/public/index.php
           [line] => 34
           [function] => run
           [class] => Zend_Application
       )

)

```

**Comment by barryo** [comment]
Hi @huddx01 

This looks a bit awkward.

> But other apache-vhosts have no problem accessing to it...

I take it you mean they have no issue accessing the LE certs for SSL? Which would be correct because Apache kicks off as root, accesses what it needs and drops to the `www-data` user then.

My inclination would be to configure sudo to that your config becomes something like:

```
defaults.mailbox.dovecot_pw_binary = "/usr/bin/sudo /usr/bin/doveadm pw"
```

and your (for example) `/etc/sudoers.d/vimbadmin` might have **(untested, read the docs)**:

```
www-data  ALL = (ALL) NOPASSWD: /usr/bin/doveadm pw *
```

**Comment by huddx01** [comment]
@barryo 

Yes, there are ne issues with LE certs for SSL... 

So, i completly renewed the Le Cert for the vimbadmin domain. After this, still didnt work...

But i followed your good Idea to add the www-data User into the sudoers for the doveadm command. 
And it works now!

Ok - its a workaround for now. Thank you so far.
But something seems to be weird...

I will reinstall the vimbadmin stuff completely. Maybe in a few days...

**Comment by mulderij** [comment]
An alternative is through php_crypt and then modify the Dovecot password query against MariaDB:

*/etc/postfixadmin/config.local.php*
```
$CONF['encrypt'] = 'php_crypt:SHA512';
```
*/etc/dovecot/dovecot-sql.conf.ext*
```
password_query = \
  SELECT \
  CASE \
    WHEN LEFT(password,4) = '$2y$' THEN CONCAT('{CRYPT}', password) \
    WHEN LEFT(password,3) = '$6$' THEN CONCAT('{SHA512-CRYPT}', password) \
    WHEN LEFT(password,3) = '$5$' THEN CONCAT('{SHA256-CRYPT}', password) \
    WHEN LEFT(password,3) = '$1$' THEN CONCAT('{MD5-CRYPT}', password) \
  END AS password \
  FROM mailbox \
  WHERE username='%u'
```
This doesn't add the the string for the encryption method in the database, but adds it in the result returned to Dovecot. For some unclear reason Dovecot needs this (the $\<code>$ should have been enough). Of course if you are only using sha512 not all rewrite cases are necessary, but they don't hurt...

**Comment by barryo** [comment]
Time out closing on this - please reopen if help still required.

**Comment by s-a-s-k-i-a** [comment]
Is this fixed in vimbadmin 3.4.x?
As I am seeing this currently in my Vimbadmin 3.3.1 installation and I have not implemented any workaround yet.
